tiledb/api/src/query_arrow/fields.rs

```diff
@@ -0,0 +1,171 @@
+
+#[derive(Debug, Default)]
+pub enum QueryField {
+    #[default]
+    Default,
+    WithCapacity(usize),
+    WithCapacityAndType(usize, adt::DataType),
+    WithType(adt::DataType),
+    Buffer(Arc<dyn aa::Array>),
+}
+
+impl QueryField {
+    pub fn capacity(&self) -> Option<usize> {
+        match self {
+            Self::Default => Some(DEFAULT_CAPACITY),
+            Self::WithCapacity(capacity) => Some(*capacity),
+            Self::WithCapacityAndType(capacity, _) => Some(*capacity),
+            Self::WithType(_) => Some(DEFAULT_CAPACITY),
+            Self::Buffer(_) => None,
+        }
+    }
+
+    pub fn target_type(&self) -> Option<adt::DataType> {
+        match self {
+            Self::Default => None,
+            Self::WithCapacity(_) => None,
+            Self::WithCapacityAndType(_, dtype) => Some(dtype.clone()),
+            Self::WithType(dtype) => Some(dtype.clone()),
+            Self::Buffer(array) => Some(array.data_type().clone()),
+        }
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct QueryFields {
+    pub fields: HashMap<String, QueryField>,
+}
+
+impl QueryFields {
+    pub fn insert<S: Into<String>>(&mut self, name: S, field: QueryField) {
+        let name: String = name.into();
+        self.fields.insert(name.clone(), field);
+    }
+}
+
+pub struct QueryFieldsBuilder {
+    fields: QueryFields,
+}
+
+impl QueryFieldsBuilder {
+    pub fn new() -> Self {
+        Self {
+            fields: Default::default(),
+        }
+    }
+
+    pub fn build(self) -> QueryFields {
+        self.fields
+    }
+
+    pub fn field(mut self, name: &str) -> Self {
+        self.fields.insert(name, Default::default());
+        self
+    }
+
+    pub fn field_with_buffer(
+        mut self,
+        name: &str,
+        buffer: Arc<dyn aa::Array>,
+    ) -> Self {
+        self.fields.insert(name, QueryField::Buffer(buffer));
+        self
+    }
+
+    pub fn field_with_capacity(mut self, name: &str, capacity: usize) -> Self {
+        self.fields.insert(name, QueryField::WithCapacity(capacity));
+        self
+    }
+
+    pub fn field_with_capacity_and_type(
+        mut self,
+        name: &str,
+        capacity: usize,
+        dtype: adt::DataType,
+    ) -> Self {
+        self.fields
+            .insert(name, QueryField::WithCapacityAndType(capacity, dtype));
+        self
+    }
+
+    pub fn field_with_type(mut self, name: &str, dtype: adt::DataType) -> Self {
+        self.fields.insert(name, QueryField::WithType(dtype));
+        self
+    }
+}
+
+pub struct QueryFieldsBuilderForQuery {
+    query_builder: QueryBuilder,
+    fields_builder: QueryFieldsBuilder,
+}
+
+impl QueryFieldsBuilderForQuery {
+    pub(crate) fn new(query_builder: QueryBuilder) -> Self {
+        Self {
+            query_builder,
+            fields_builder: QueryFieldsBuilder::new(),
+        }
+    }
+
+    pub fn end_fields(self) -> QueryBuilder {
+        self.query_builder.with_fields(self.fields_builder.build())
```

Comment:
If a field is specified in the first `QueryBuilder` then this will replace it with the new request right?

Reply:
Yep, the idea was that users would use one or the other, not both. I really think the field spec needs work because of the way it limits easily adding options. Off the top of my head, perhaps something like such:

```rust
    let buffer = Arc::new(Int32Array::from(vec![0u32; 10]));
    let mut query = QueryBuilder::new()
        .with_fields([
            ("example_1",  buffer),
            ("example_2", vec![FieldArg::Capacity(100)]),
            ("example_3", vec![FieldArg::Capacity(100), FieldArg::RelaxedConversion])
        ])
        .build()?;
```

And then `with_fields` would just take a `Into<QueryFields>` generic type. I suddenly miss atoms for API design...